### ISSUE TITLE
Restic backup of root of a drive letter on windows refuses to restore properly

### ISSUE THREAD

**Issue by addos** [restic/restic#2004]
Output of `restic version`
--------------------------
restic 0.9.2 compiled with go1.10.3 on windows/amd64

How did you run restic exactly?
-------------------------------
restic -r D:\backups backup E:\
restic -r D:\backups restore <id> -t target_dir

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

Expected behavior
-----------------
It should be able to store/restore properly

Actual behavior
---------------
Restic apparently stores : in the paths, which make the repo unable to be restored?

When I run ls against my repo, it shows something like the following:
/E/E:./repo/.git/objects/info
/E/E:./repo/.git/objects/pack
/E/E:./repo/.git/refs
/E/E:./repo/.git/refs/heads
/E/E:./repo/.git/refs/heads/git-annex
/E/E:./repo/.git/refs/heads/synced
/E/E:./repo/.git/refs/heads/synced/git-annex
/E/E:./repo/.git/refs/heads/synced/master
/E/E:./repo/.git/refs/tags

When I try to restore, it shows something like the following:
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\objects\f6: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\objects\f6\65cfca35e159a0a47237041d1d20edd0218ba3: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\objects\info: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\objects\pack: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\heads: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\heads\git-annex: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\heads\synced: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\heads\synced\git-annex: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\heads\synced\master: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
ignoring error for C:\Users\brad\a\E\E:.\repo\.git\refs\tags: mkdir C:\Users\brad\a\E\E:.: The filename, directory name, or volume label syntax is incorrect.
There were 371 errors

Steps to reproduce the behavior
-------------------------------
Basically, backup the root of a drive letter on windows, and then try to restore it.

For example:
restic -r repo backup E:\

Do you have any idea what may have caused this?
-----------------------------------------------
I think restic is storing colons in the paths, which it can't properly restore on windows.

Do you have an idea how to solve the issue?
-------------------------------------------
I think restic is storing colons in the paths, which it can't properly restore on windows.

Did restic help you or made you happy in any way?
-------------------------------------------------
All the darn time, I absolutely love this project.

**Comment by jirib** [comment]
I suppose this disqualifies restic to be able to restore backup of Windows C:\ drive to another drive while booted from WinPE media for disaster recovery purposes - that is, when you boot WinPE from eg. usb disk, your C:\ would be for example E:\ and you would need to restore C:\ from backup to E:\.

Am I right?

**Comment by fd0** [comment]
> Am I right?

Not quite: You can do that with restic, even with this limitation. You only need to make sure that the backup is made by changing the current working directory to `c:\` and calling restic like this: `restic backup .`. The top-level files/folders in the snapshot will then be the content of `c:\`, without the drive letter. You can just restore that to `c:\` again, no problem.

If you run `restic backup c:\` instead, that'll make restic use the drive letter as the top-level directory, so `c:\boot.ini` will be archived as `/c:/boot.init`, and restic currently has no option to strip the additional directory `c:` directory for restore.

**Comment by Swonkie** [comment]
It seems to me that, apart from the colon issue (which I didn't see), the way paths are stored does not allow to restore **anything** properly on Windows if an absolute was used to make the backup. I backed up `C:\User\Swonkie` and wasn't able to restore files in their original place. Restic would always create a folder structure like `C:\C\Users\Swonkie\...` (notice the folder called `C`). So basically it's a must to `cd` into the directory to be backed up and use `.`, which is very cumbersome.

**Comment by tulsti** [comment]
One may backup files without using a drive letter, so just use `\users\swonkie` and you get rid of the drive letter problem. This works just bad when having multiple drives.
It also may be a feature to strip parts of the path during restore, so making it possible not to use c: and avoid such problems.

**Comment by Swonkie** [comment]
I see, @tulsti 
This is still inconvenient though, because I need to ensure I'm on the correct drive before using the relative path.

What is the thought behind storing the path in front of the backed up folder anyway? Wouldn't it be easiest and most flexible to just store the folder `Swonkie` when backing up `C:\Users\Swonkie`. When restoring, I could then say target: `C:\Users` or `\Users` or `.`

What do we gain from storing the path?

**Comment by BrianMBeaulieu** [comment]
I'm working on implementing restic across multiple platforms and for Windows, we need to back up D:\Folder .. so drive letter support is needed especially in server environments.

**Comment by cfbao** [comment]
@Swonkie Not storing the full path could be problematic when backing up multiple folders into the repository.
For example, `restic backup C:\folder D:\folder`. Without full path, you can then expect a lot of collisions upon restore.

**Comment by tulsti** [comment]
One may just think of a non technical solution. Instead of just putting everything into one repo you can split it up. We one folder per server and subfolders for the drives containing the repos.
So no need to store drive letters..
Am 15.11.18, 15:33, Brian Beaulieu <notifications@github.com> schrieb:
I'm working on implementing restic across multiple platforms and for Windows, we need to back up D:\Folder .. so drive letter support is needed especially in server environments.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/restic/restic","title":"restic/restic","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/restic/restic"}},"updates":{"snippets":[{"icon":"PERSON","message":"@BrianMBeaulieu in #2004: I'm working on implementing restic across multiple platforms and for Windows, we need to back up D:\\Folder .. so drive letter support is needed especially in server environments."}],"action":{"name":"View Issue","url":"https://github.com/restic/restic/issues/2004#issuecomment-439060025"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/restic/restic/issues/2004#issuecomment-439060025",
"url": "https://github.com/restic/restic/issues/2004#issuecomment-439060025",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [restic/restic] Restic backup of root of a drive letter on windows refuses to restore properly (#2004)",
"sections": [
{
"text": "",
"activityTitle": "**Brian Beaulieu**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@BrianMBeaulieu",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"restic/restic\",\n\"issueId\": 2004,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close issue",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueClose\",\n\"repositoryFullName\": \"restic/restic\",\n\"issueId\": 2004\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/restic/restic/issues/2004#issuecomment-439060025"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 383067656\n}"
}
],
"themeColor": "26292E"
}
]

**Comment by silicontrip** [comment]
> One may backup files without using a drive letter, so just use `\users\swonkie` and you get rid of the drive letter problem. This works just bad when having multiple drives.
> It also may be a feature to strip parts of the path during restore, so making it possible not to use c: and avoid such problems.

I was unable to use \ or even . as restic (0.12.0) adds the C:\ to the path.
I second the request for a strip path option, similar to `patch`'s -p option.

**Comment by nathanchere** [comment]
+1, everything about restic has been great except for this which is not an absolute deal-breaker but it's far from ideal. I would hope to be able to back up and restore with a relative path rather than the full path being enforced all the time.

**Comment by nathanchere** [comment]
One possible implementation I'd like to see. Keeping in mind snapshots like the following:

```
ID        Time                 Host             Tags        Paths
---------------------------------------------------------------------------------------
abcd1234  2000-01-01 23:59:59  TestVm24822              C:\foo
---------------------------------------------------------------------------------------
1 snapshots
```

and let's assume the following file system contents:

```
    Directory: C:\temp\emailtemplates

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a---           1/1/2000 11:59 PM           1024 example.txt
```

instead of using `--target` (which for Windows hosts it problematic as per this issue) I am thinking a flag `--relative-target {path}` which  allows you to specify the destination path as normal but strips the original path prefix from all output.
 
For example if use `restic restore abcd1234 --target E:\foo` I will instead end up with `E:\foo\c\foo\example.txt`. With 
`restic restore abcd1234 --relative-target E:\foo` I would expect the `c:\foo` path to no longer be included in the output path so as to get `E:\foo\example.txt`.

**Comment by rhoellwarth** [comment]
Moin!

For those who have a backup with the drive-letter in the path e.g. /P/P:./Documents/ and NEED to restore it - this is a working solution:
* Install Git for Windows (it comes with Bash and tar)
* start the (Git) Bash and navigate to a temporary space - cd /c/Windows changes e.g. to C:\Windows or cd $TEMP etc.
* call restic dump and pipe it through tar: "restic dump -a tar 1234asdf '/' | tar -xvf -"
* as the bash CAN create foldernames like 'P:.' tar will just output the whole ('/' = all) Snapshot 1234asdf
* rename the problematic folder: "mv 'P:.' P"
* the files are properly restored and the offending foldername is gone - they can be copied/moved to their destination path

To fix that problem in restic itself I suggest an option --strip-path 'string' that just removes 'string' from the beginning of every path that should be restored. Entries that don't start with this string should be warned about and NOT restored as they might be overwriting files that had 'string' removed from their path. On Windows one could then just use --strip-path '/P/P:./' (as to my example above) and all should be well. Additionally there should perhaps be a warning when restic creates backups with driveletters in the saved filename, hinting to the option --strip-path from the restore command.

greetings from Germany
ronald

**Comment by ffalcon31415** [comment]
This bug is open since 2018 but isn't assigned and seems to be orphaned.
I'm a Windows user and I backup up a root drive (because nobody told me not to) to B2. 
I tried a test restore on Windows today for the first time (previously have successfully restored to linux, which obviously didn't reveal the issue).

I'm shocked that I can't restore my backups to windows.

I tried @rhoellwarth workaround but didn't work for me -- command not found for 'restic' in git bash.

Is there any prospect of this being fixed? It should be added to the documentation as it's a major bug for Windows users, especially if we're not being warned about it before making the backup.

**Comment by AndrewSav** [comment]
> This bug is open since 2018 but isn't assigned and seems to be orphaned.

Just to clarify, this is an open-source non-profit project run by volunteers, it is provided "as is" without any warranties, including but not limited to fitness for a particular purpose.

If you are new to open source scene, this is basically how it works: you find a bug, you create an issue here on the tracker. If some one wants to work on it (this includes yourself as well), they can leave a note here discussing approach, and if agreed by the maintainers draft a PR, get it reviewed and finally merged. The vast majority of open source project do not "assign" issues, instead they are just getting picked up (or not). When they are "assigned" most of the time, it's for the maintainers convenience to indicate that someone is working on they issue, it is not a commitment to you, that it is going to be fixed.

> I'm shocked that I can't restore my backups to windows.

I'd take it as a lesson for all of us to always test our mission-critical workflows. It would be pretty bad if you backed up something and then cannot restore it just because you have not tested it with your use case before. If it is not mission-critical though, it's probably not a big deal. Software has bugs occasionally and even if it does not, one can always make a mistake or a wrong assumption in configuring it. I personally would not rest peacefully, until I've tested that I can actually restore my backup.

> I tried @rhoellwarth workaround but didn't work for me -- command not found for 'restic' in git bash.

I suggest persevering and googling for a solution here in the meanwhile. The error looks like it indicates that the 'restic' command cannot be found, so you need to check the usual things to resolve it (look up how shell looks for executables to run when you issue a command, you should be able either specify path to 'restic' on the command line, or add it to `PATH`)

> Is there any prospect of this being fixed?

It is not likely to get the answer you wish for. It will be fixed when someone is motivated enough to work on, has a plan how to fix it and gets buy-in from the maintainers that this is the way to go. When or if that happens, I'm afraid no one can tell you unless someone already has such plans.

**Comment by rawtaz** [comment]
@ffalcon31415 As @AndrewSav hinted, the "command not found for 'restic' in git bash" suggests that you haven't got `restic` in the PATH of the shell, or similar. Maybe it's named `restic.exe`? Regardless, try providing the full path to the restic binary, wherever it's located, when you run it. E.g. `C:\Users\ffalcon31415\Downloads\restic.exe`.

**Comment by ffalcon31415** [comment]
@AndrewSav 
Thank you for the suggestions. 
I know there's no warranty -- that's why I'm not suing! jk
I'm not familiar with the language around "assigning" etc. of issues except as I've seen them referred to on some larger projects. Apologies for the tone of my comment above.
I do have access to my backups (through linux).

Fair enough over all re your comments. I do feel that the docs could contain a one-liner warning of this issue, even if nobody has time/inclination to find and implement a fix for the bug. I'll see if I can submit one though that's outside of my experience (never having contributed to open source software (or any software aside from my own scripts). 

@rawtaz and @AndrewSav re git bash issue
Thanks that makes sense and will give it a go.

**Comment by unshuffled** [comment]
Resurrecting this old post to provide a workaround for those poor suckers like me who find ourselves with a backup we can't restore because of the drive letter bug. This fix only works if you have admin access to the Windows machine involved.

I was able to create a Symbolic Link called "C" (for my C drive backup) that I pointed to the root directory of the drive using a Power Shell run as Administrator:
`cd C:\ ; 
New-Item -ItemType SymbolicLink -Path "C" -Target "."`

I could then restore everything by setting the restic target to "\".

I think going forward I will use relative paths and run separate restic backups for each drive. I'll tag these backups "c_drive", "d_drive", etc. for disambiguation.

**Comment by unshuffled** [comment]
> 

Weirdly, restic shows the full drive letters when using "restic snapshots" even when backups do not specify the full drive letters. This raised my blood pressure for a moment. Looking closer, "restic ls" shows the file paths without the drive letters, so I think this is working alright.

I'm wondering if using tags for the drive letters is sufficient disambiguating. I also wonder if this impacts the deduplication process negatively in any way.

**Comment by hiinaspace** [comment]
I tripped over this rake today, where files backed up using `restic backup H:` show up in `restic ls` as `/H/H:./some/file/path/`. Running `restore` then tries to make the `H:.` directory and fails with "The filename, directory name, or volume label syntax is incorrect". Pretty unfortunate. I agree this should definitely be called out in the docs.

In this case I only needed to restore a few files, so I offer another workaround to those that come across this bug: `restic dump --archive zip /H/H:./some/file/path > H:/restore.zip` will make a zip file with the files in the directory, which you can extract with another program (e.g. 7-zip), sidestepping the colons in the path issue. It'd definitely be nicer if `restic restore` could do this too though, even if it is some kludge like replacing `:` with `_` or similar.

**Comment by MichaelEischer** [comment]
`restic restore latest:/H/H:. -t target` should also work. That uses the `<snapshot>:<subpath>` syntax to skip the problematic part.

**Comment by durcom2** [comment]
Another issue related to this...

Using restic on windows to back up c: drive.  I used the workaround suggested by fd0 above...cd c:\   then run restic with source folder as  \

Works correctly, however file excludes do NOT.

Specifically, if I try to exclude System Volume Information folder, in the exclude file, specifying \System Volume Information or
/System Volume Information does NOT work.

Specifying just System Volume Information with no \ DOES work, but then it will also exclude that folder from any subfolders as well. In order to get it to work properly, I have to specify *:\System Volume Information in the exclude file which does work properly.  But, kind of goes against the recommendation of not using drive letters with windows...Should be consistent...
